libs/contract-core/contracts/TheCoin.sol

```diff
@@ -22,281 +17,52 @@ import "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
-
-    function setMinter(address newMinter) public
-    onlyMinter
-    {
-        new_Minter = newMinter;
-    }
-    function acceptMinter() public
-    {
-        require(msg.sender == new_Minter, "Permission Denied");
-        role_Minter = new_Minter;
-        new_Minter = address(0);
-    }
-
-    function setTheCoin(address newCoinManager) public
-    onlyTheCoin
-    {
-        new_TheCoin = newCoinManager;
-    }
-    function acceptTheCoin() public
-    {
-        require(msg.sender == new_TheCoin, "Permission Denied");
-        role_TheCoin = new_TheCoin;
-        new_TheCoin = address(0);
-    }
-
-    function setPolice(address newPolice) public
-    onlyPolice
-    {
-        new_Police = newPolice;
-    }
-    function acceptPolice() public
-    {
-        require(msg.sender == new_Police, "Permission Denied");
-        role_Police = new_Police;
-        new_Police = address(0);
-    }
-
-    // ------------------------------------------------------------------------
-    // Owner can transfer out any ERC20 tokens accidentally assigned to this contracts address
-    // ------------------------------------------------------------------------
-    function transferAnyERC20Token(address tokenAddress, uint256 tokens) public onlyTheCoin returns (bool success) {
-        IERC20Upgradeable tokenContract = IERC20Upgradeable(tokenAddress);
-        return tokenContract.transfer(role_TheCoin, tokens);
-    }
-
-    // ------------------------------------------------------------------------
-    // Override hooks to ensure isTransferable is called for every transfer
-    // ------------------------------------------------------------------------
-    function _beforeTokenTransfer(address from, address to, uint256 amount) isTransferable(from, amount)
-        internal virtual override
-    {
-      super._beforeTokenTransfer(from, to, amount);
-    }
-
-    ///////////////////////////////
-    // modifier balanceAvailable(uint amount) {
-    //     require(_balances[msg.sender] >= amount, "Caller has insufficient balance");
-    //     _;
-    // }
-
-    modifier isTransferable(address from, uint amount) {
-        //require(_balances[from] >= amount, "Caller has insufficient balance");
-        require(freezeUntil[from] < block.timestamp, "Caller's account is currently frozen");
-        _;
-    }
-
-    modifier onlyMinter()
-    {
-        require(msg.sender == role_Minter, "Invalid sender");
-        _;
-    }
-
-    modifier onlyTheCoin()
-    {
-        require(msg.sender == role_TheCoin, "Invalid sender");
-        _;
-    }
-
-    modifier onlyPolice()
-    {
-        require(msg.sender == role_Police, "Invalid sender");
-        _;
-    }
-
-    modifier timestampIncreases(address from, uint256 timestamp)
-    {
-        require(timestamp > lastTxTimestamp[from], "Provided timestamp is lower than recorded timestamp");
-        _;
-    }
+contract TheCoin is Pluggable {
```

Comment:
I broke down this file into a series of smaller contracts to isolate each piece of functionality into a more testable unit.

Also, re-wrote most of it to support plugins, use roles, and dropped tap-cap-manager